android/app/src/main/java/com/zulipmobile/sharing/ReceiveShareActivity.kt

```diff
@@ -0,0 +1,129 @@
+package com.zulipmobile.sharing
+
+import android.content.Context
+import android.content.Intent
+import android.content.res.Configuration
+import android.net.Uri
+import android.os.Bundle
+import android.os.Parcelable
+import android.util.Log
+import android.webkit.WebView
+import androidx.annotation.Nullable
+import com.facebook.react.ReactActivity
+import com.facebook.react.ReactApplication
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.bridge.WritableMap
+import com.facebook.react.bridge.Arguments
+import com.zulipmobile.MainActivity
+import com.zulipmobile.notifications.*
+
+const val TAG = "ZulipReceiveShare"
+
+class ReceiveShareActivity : ReactActivity() {
+
+    /**
+     * Returns the name of the main component registered from JavaScript.
+     * This is used to schedule rendering of the component.
+     */
+    override fun getMainComponentName(): String? {
+        return "SharingRoot"
+    }
+
+    private fun sendEvent(reactContext: ReactContext,
+                          eventName: String,
+                          @Nullable params: WritableMap) {
+        Log.d(TAG, "Sending event with shared data")
+        emit(reactContext, eventName, params)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        WebView.setWebContentsDebuggingEnabled(true)
+        val application = application as ReactApplication
+        launchMainActivity(application as Context)
+        val host = application.reactNativeHost
+        val reactContext = host.tryGetReactInstanceManager()?.currentReactContext
+        val intent = intent
+        when {
+            intent?.action == Intent.ACTION_SEND -> {
+                when {
+                    "text/plain" == intent.type -> handleSendText(reactContext, intent)
+                    intent.type?.startsWith("image/") == true -> handleSendImage(reactContext, intent)
+                    else -> handleGenericFile(reactContext, intent)
+                }
+            }
+        }
+        finish()
+    }
+
+    private fun handleSendText(reactContext: ReactContext?, intent: Intent) {
+        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
+        val params = Arguments.createMap()
+        params.putString("type", "text")
+        params.putString("sharedText", sharedText)
+        if (sharedText != null) {
+            if (reactContext == null) {
+                SharingModule.initialSharedData = params
+            } else {
+                sendEvent(reactContext, "shareReceived", params)
+            }
+
+        }
+    }
+
+    private fun handleSendImage(reactContext: ReactContext?, intent: Intent) {
+        (intent.getParcelableExtra<Parcelable>(Intent.EXTRA_STREAM) as? Uri)?.let {
```

Comment:
Looks like the version in this revision is instead:
```kt
        val uri = intent.getParcelableExtra(Intent.EXTRA_STREAM) as? Uri ?: return null
```
with only the cast `as? Uri`, and no type parameter `<Uri>`.

I wasn't sure what that would do, so I hit Ctrl+Shift+P to get the type of the expression `intent.getParcelableExtra(Intent.EXTRA_STREAM)`. That informed me the type was `Uri!` (which means "`Uri` or `Uri?`, sorry it comes from a Java API and the Java type system doesn't express that distinction".) So I *think* that means the type parameter gets inferred as `Uri`, and that gets passed through, so that the cast inside the implementation is to `Uri`, and our own cast doesn't actually do anything at runtime.

Given that that's the case, probably best to be explicit and have the code match what's really going on:
```kt
        val uri = intent.getParcelableExtra<Uri>(Intent.EXTRA_STREAM) ?: return null
```

Reply:
Done. But instead of returning `null`, it throws an exception now.